design/open/staging-compaction.md

```diff
@@ -0,0 +1,177 @@
+atomically update the branch entity:
+
+1. Set the `CompactedMetaRangeID` to the new MetaRangeID.
+1. Remove the compacted `SealedTokens` from the branch.
+
+The Compactor should fail the operation if the sealed tokens have changed 
+since the compaction started (SetIf). Although the algorithm can tolerate 
+additional sealed tokens being added during the compaction, it's better to 
+avoid competing with concurrent commits. Commits have the same benefits as 
+compactions, but they are proactively triggered by the user (and might fail 
+if compaction succeeds).
+
+Consistency guarantees for the compaction process are derived from the 
+Commit operation consistency guarantees. In short, the compaction process 
+follows the same principles as the Commit operation. Writes to a branch 
+staging token succeeds only if it's the staging token after the write 
+occurred. Therefore, replacing the sealed token with an equivelant MetaRange 
+guarantees no successful writes are gone missing.
+
+### Branch Write Flow
+
+No change here. Writes are performed on the staging token.
+
+### Branch Read Flow
+
+Reading from a branch includes both a specific entry lookup and a listing 
+request on entries of a branch. The read flow is more complicated than what 
+we have today, and is different for compacted and uncompacted branches.
+
+#### Combined (Committed+ Uncommitted)
+
+There are 3 layers of reading from a branch:
+1. Read from the staging token.
+2. Read from the sealed tokens (in order).
+3. Depends on compaction:
+   1. If a CompactedMetaRangeID exists, read from the compacted MetaRange.
+   1. If a CompactedMetaRangeID doesn't exist, read from the CommitID.
+
+#### Committed
+
+Just like today, reads are performed on the Branch's CommitID.
+
+#### Uncommitted
+
+For operations such as `DiffUncommitted` or checking if the staging 
+area is empty, the read flow will be as follows:
+
+1. Read from the staging token.
+2. Read from the sealed tokens (in order).
+3. If a CompactedMetaRangeID exists, read the 2-way diff between the compacted
+   metarange and the CommitID's metarange.
+
+* There's an inefficiency here, as there's an option we'll need to read 2 whole 
+  metaranges to get the diff, like when there's a single change in every 
+  range. The nature of changes to a lakeFS branch is such that changes are 
+  expected in a small number of ranges, and the diff operation is expected 
+  to skip most ranges. Moreover, Pebble caching the ranges should remove the 
+  most costly operation of ranges comparison - fetching them from S3. If 
+  this is still inefficient, we can use the immutability trait[^3] of the 
+  diff: we can calculate the diff result once and cache it.
+
+### Commit Flow
+
+The commit flow is slightly affected by the compaction process. If 
+compaction never happened, the commit flow is the same as today. If a 
+compaction happened, apply the changes to the compacted metarange instead of 
```

Comment:
What if a compaction happened during a commit? will this fail the commit? If that is the case, are we fine with compaction failing commits?

Reply:
Updated.